vpr/src/base/xdc_constraints.cpp

```diff
@@ -0,0 +1,244 @@
+#include <istream>
+#include <string>
+#include <sstream>
+#include <fstream>
+#include <utility>
+#include <functional>
+#include <tcl/tcl.h>
+
+#include "globals.h"
+#include "vpr_constraints.h"
+#include "atom_netlist.h"
+
+#include "oo_tcl.h"
+#include "xdc_constraints.h"
+
+
+enum class e_XDCProperty {
+    XDC_PROP_PACKAGE_PIN,
+    XDC_PROP_IOSTANDARD,
+    XDC_PROP_UNKNOWN
+};
+
+static e_XDCProperty xdc_prop_from_str(const char* str) {
+    if (!std::strcmp(str, "PACKAGE_PIN"))
+        return e_XDCProperty::XDC_PROP_PACKAGE_PIN;
+    if (!std::strcmp(str, "IOSTANDARD"))
+        return e_XDCProperty::XDC_PROP_IOSTANDARD;
+    
+    return e_XDCProperty::XDC_PROP_UNKNOWN;
+}
+
+DECLARE_TCL_TYPE(AtomPortId, port_tcl_t)
+
+class TclPhysicalConstraintsClient : public TclClient {
+public:
+    VprConstraints& constraints;
+    const t_arch& arch;
+    AtomNetlist& netlist;
+
+    TclPhysicalConstraintsClient(VprConstraints& constraints_, const t_arch& arch_, AtomNetlist& netlist_) :
+        constraints(constraints_),
+        arch(arch_),
+        netlist(netlist_)
+    {}
+
+    typedef int(TclPhysicalConstraintsClient::*TclMethod)(int objc, Tcl_Obj* const objvp[]);
+    void register_methods(std::function<void(const char* name, TclMethod)> register_method) {
+        register_method("get_ports", &TclPhysicalConstraintsClient::get_ports);
+        register_method("set_property", &TclPhysicalConstraintsClient::set_property);
+    }
+
+    /* TCL functions */
+
+    int set_property(int objc, Tcl_Obj* const objvp[]) {        
+        if (objc < 3)
+            return this->_ret_error("set_property: Expected at least 2 arguments, got " + 
+                                    std::to_string(objc) + ".");
+
+        const char* property_name = Tcl_GetString(objvp[1]);
+        if (property_name == nullptr)
+            return this->_ret_error("set_property: First argument should be a string.");
+        
+        e_XDCProperty property = xdc_prop_from_str(property_name);
+
+        switch (property) {
+        case e_XDCProperty::XDC_PROP_PACKAGE_PIN:
+            if (objc != 4)
+                return this->_ret_error("set_property: Property `PACKAGE_PIN` "
+                                        "requires one target and one value.");
+            return this->_set_property_package_pin(objvp[2], objvp[3]);
+        case e_XDCProperty::XDC_PROP_IOSTANDARD:
+            if (objc != 4)
+                return this->_ret_error("set_property: Property `IOSTANDARD` "
+                                        "requires one value and one target.");
+            return this->_set_property_iostandard(objvp[2], objvp[3]);
+        case e_XDCProperty::XDC_PROP_UNKNOWN:
+            return this->_ret_error("set_property: Property `" + std::string(property_name) +
+                            "` is not recognized.");
+        default: break;
```

Comment:
This should probably be indented. You can run `make format` which will use the autoformatter.

Reply:
I might be wrong, but formatting may still need to be performed here